File: java/com/google/copybara/http/auth/UsernamePasswordInterceptor.java
/*
 * Copyright (C) 2023 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.copybara.http.auth;

import com.google.api.client.http.HttpExecuteInterceptor;
import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableSetMultimap;
import com.google.copybara.credentials.CredentialIssuingException;
import com.google.copybara.credentials.CredentialModule.UsernamePasswordIssuer;
import com.google.copybara.credentials.CredentialRetrievalException;

/** Representation for authentication information for an http request. */
public class UsernamePasswordInterceptor implements AuthInterceptor {

  private final UsernamePasswordIssuer issuer;

  public UsernamePasswordInterceptor(UsernamePasswordIssuer issuer) {
    this.issuer = Preconditions.checkNotNull(issuer);
  }

  @Override
  public HttpExecuteInterceptor interceptor() throws
      CredentialRetrievalException, CredentialIssuingException {
    String un = issuer.username().issue().provideSecret();
    String pwd = issuer.password().issue().provideSecret();
    return (req) -> req.getHeaders().setBasicAuthentication(un, pwd);
  }

  @Override
  public ImmutableList<ImmutableSetMultimap<String, String>> describeCredentials() {
    return issuer.describeCredentials();
  }
}

File: src/Copybara.Core/Http/Auth/UsernamePasswordInterceptor.cs
/*
 * Copyright (C) 2023 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System.Linq;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using Copybara.Common;
using Copybara.Credentials;

namespace Copybara.Http.Auth;

/// <summary>Representation for authentication information for an http request.</summary>
public class UsernamePasswordInterceptor : IAuthInterceptor
{
    private readonly UsernamePasswordIssuer _issuer;

    public UsernamePasswordInterceptor(UsernamePasswordIssuer issuer)
    {
        _issuer = Preconditions.CheckNotNull(issuer);
    }

    public Action<HttpRequestMessage> Interceptor()
    {
        string un = _issuer.Username.Issue().ProvideSecret();
        string pwd = _issuer.Password.Issue().ProvideSecret();
        string encoded = Convert.ToBase64String(Encoding.UTF8.GetBytes($"{un}:{pwd}"));
        return req => req.Headers.Authorization = new AuthenticationHeaderValue("Basic", encoded);
    }

    public IReadOnlyList<ImmutableListMultimap<string, string>> DescribeCredentials() =>
        _issuer.DescribeCredentials()
            .Select(MultimapConversions.ToListMultimap)
            .ToList();
}
